qutebrowser/browser/mhtml.py

```diff
@@ -0,0 +1,534 @@
+        """
+        if self._used:
+            raise ValueError("Downloader already used")
+        self._used = True
+        web_url = self.web_view.url()
+        web_frame = self.web_view.page().mainFrame()
+
+        self.writer = MHTMLWriter(
+            web_frame.toHtml().encode('utf-8'),
+            content_location=urlutils.encoded_url(web_url),
+            # I've found no way of getting the content type of a QWebView, but
+            # since we're using .toHtml, it's probably safe to say that the
+            # content-type is HTML
+            content_type='text/html; charset="UTF-8"',
+        )
+        # Currently only downloading <link> (stylesheets), <script>
+        # (javascript) and <img> (image) elements.
+        elements = web_frame.findAllElements('link, script, img')
+
+        for element in elements:
+            element = webelem.WebElementWrapper(element)
+            # Websites are free to set whatever rel=... attribute they want.
+            # We just care about stylesheets and icons.
+            if not _check_rel(element):
+                continue
+            if 'src' in element:
+                element_url = element['src']
+            elif 'href' in element:
+                element_url = element['href']
+            else:
+                # Might be a local <script> tag or something else
+                continue
+            absolute_url = web_url.resolved(QUrl(element_url))
+            self.fetch_url(absolute_url)
+
+        styles = web_frame.findAllElements('style')
+        for style in styles:
+            style = webelem.WebElementWrapper(style)
+            if 'type' in style and style['type'] != 'text/css':
+                continue
+            for element_url in _get_css_imports(str(style)):
+                self.fetch_url(web_url.resolved(QUrl(element_url)))
+
+        # Search for references in inline styles
+        for element in web_frame.findAllElements('[style]'):
+            element = webelem.WebElementWrapper(element)
+            style = element['style']
+            for element_url in _get_css_imports(style, inline=True):
+                self.fetch_url(web_url.resolved(QUrl(element_url)))
+
+        # Shortcut if no assets need to be downloaded, otherwise the file would
+        # never be saved. Also might happen if the downloads are fast enough to
+        # complete before connecting their finished signal.
+        self.collect_zombies()
+        if not self.pending_downloads and not self._finished:
+            self.finish_file()
+
+    def fetch_url(self, url):
+        """Download the given url and add the file to the collection.
+
+        Args:
+            url: The file to download as QUrl.
+        """
+        if url.scheme() not in {'http', 'https'}:
+            return
+        # Prevent loading an asset twice
+        if url in self.loaded_urls:
+            return
+        self.loaded_urls.add(url)
+
+        log.downloads.debug("loading asset at %s", url)
```

Comment:
Please use `str.format` even for logging - it's usually considered bad practice because of the performance hit when you don't need to log, but we're logging everything to RAM anyways. So consistency wins :wink:

I've noticed that in some other places, I won't comment on them separately.